core/src/main/scala/cats/data/ContT.scala

```diff
@@ -0,0 +1,109 @@
+package cats
+package data
+
+/**
+ * This is a continuation transformer based on the ContT in
+ * the Haskell package Control.Monad.Cont
+ *
+ * This is reasonably straight-forward except that to make
+ * a tailRecM implementation we leverage the Defer type class to
+ * obtain stack-safety.
+ */
+sealed abstract class ContT[M[_], A, B] extends Serializable { self =>
+  final def run: (B => M[A]) => M[A] = runAndThen
+  protected def runAndThen: AndThen[B => M[A], M[A]]
+
+  def map[C](fn: B => C)(implicit M: Defer[M]): ContT[M, A, C] = {
+    // allocate/pattern match once
+    val fnAndThen = AndThen(fn)
+    ContT { fn2 =>
+      val cb = fnAndThen.andThen(fn2)
+      M.defer(run(cb))
+    }
+  }
+
+  /**
+   * c.mapCont(f).run(g) == f(c.run(g))
+   */
+  def mapCont(fn: M[A] => M[A]): ContT[M, A, B] =
+    // Use later here to avoid forcing run
+    ContT.later(runAndThen.andThen(fn))
+
+  /**
+   * cont.withCont(f).run(cb) == cont.run(f(cb))
+   */
+  def withCont[C](fn: (C => M[A]) => B => M[A]): ContT[M, A, C] =
+    // lazy to avoid forcing run
+    ContT.later(AndThen(fn).andThen(runAndThen))
+
+  def flatMap[C](fn: B => ContT[M, A, C])(implicit M: Defer[M]): ContT[M, A, C] = {
+    // allocate/pattern match once
+    val fnAndThen = AndThen(fn)
+    ContT[M, A, C] { fn2 =>
+      val contRun: ContT[M, A, C] => M[A] = (_.run(fn2))
+      val fn3: B => M[A] = fnAndThen.andThen(contRun)
+      M.defer(run(fn3))
```

Comment:
I stand very much corrected here. **I'm convinced that this *is* stack-safe.** I owe you an apology, @johnynek. Dear internets, please link here if you're looking for a handy example of me being wrong.

When I first reviewed this PR on my phone, it looked like a slightly different version of the trick you used in #1400, but this is considerably better. You're basically inheriting the stack-safety of the underlying monad by explicitly hitting a stack-safe join at this point and returning to its trampoline (which it must have due to `Defer`). I'm not sure you even need the `AndThen`, since the stack is cut by the `defer`.

Reply:
We all make mistakes here and there. I very much admire folks who feel no shame about it. Kudos to you.